cmd/rootMain.go

```diff
@@ -59,6 +59,25 @@ func rootMain(cmd *cobra.Command, args []string) exitCode {
 	// Set plugin interactivity to false if execfile or rootCommandFlag were specified.
 	plugin.InitInteractive(execfile == "" && rootCommandFlag == "")
 
+	// If interactive and this process is in its own process group, then fork and run the original
+	// command as a new process that's not the leader of its process group. This is specifically to
+	// work with shells that try to use the parent's process group, which we change later on. By
```

Comment:
What does "try to use the parent's process group" mean? That some shells try to use the parent's process group's TTY?

Reply:
In this case I think it's returning the TTY to the parent process group. But I don't fully understand what zsh is doing.

Reply:
Ok. Now say `P` is the server process in pgrp `A`, `C` is the child shell process in pgrp `B`, and that `C` needs `A` (it is one of those shells you mention in the comment). Without this code, we'd get `P -- B` at the end, which resulted a failure because it didn't match `A`. Does that mean that `C` knows it needs `A` when it boots up (i.e. zsh statically determines the parent's pgrp instead of checking it via a system call)?

It's OK if the answer to this is "I have no idea." Just trying to understand it.

Reply:
I suspect it's actually: `C` asks for its parent process id (`P`) and requests the process group (`A`) for that. It then calls `tcsetpgrp` or something similar in `ioctl` (https://github.com/zsh-users/zsh/blob/ad9f07e66fded8a44adba15d576960cef587f9d4/Src/utils.c#L4643-L4674) to transfer the TTY to that process group. Originally that would've worked because `pgrp must be a (nonempty) process group belonging to the same session as the calling process` (https://linux.die.net/man/3/tcsetpgrp), but now `A` is actually a different process group that's no longer part of the same session as `C` (I think `setsid` does that).

Re-reading https://blog.nelhage.com/2010/01/a-brief-introduction-to-termios-signaling-and-job-control/, it's possible that there's an additional step I need to take after moving the daemon to acquire a controlling terminal.

Reply:
I'm going to keep digging into this, because I think there is a little more setup missing. For example
```
USER           PID  PPID  PGID   SESS JOBC STAT   TT       TIME COMMAND
michaelsmith 25735 25734 25735      0    0 Ss     ??    0:04.33 wash
michaelsmith 25167 25166 25167      0    1 S    s001    0:00.86 -zsh
michaelsmith 25734 25167 25734      0    1 S    s001    0:00.09 wash
michaelsmith 25745 25735 25745      0    0 S+   s001    0:00.18 /bin/zsh
michaelsmith 26275 26274 26275      0    1 S    s003    0:00.71 -zsh
michaelsmith 26436 26275 26436      0    1 S    s003    0:00.12 zsh
michaelsmith 26987 26436 26987      0    1 S+   s003    0:00.11 zsh
```

This shows two process trees:
- Run `wash`: 25167 (login zsh) -> 25734 (wash) -> 25735 (sub-wash) -> 25745 (/bin/zsh)
- Run `zsh`, then `zsh` again: 26275 (login zsh) -> 26436 (zsh) -> 26987 (zsh)

Digging into `ps` (on macOS), here are some notable traits:
- everything in a login shell shares the same TTY (except sub-wash, which doesn't seem to know its TTY)
- everything has a session of 0. This seems weird, I think ps on macOS is just not capturing anything useful there.
- `/bin/zsh` has a job control count of 0; seems like it should be 1
- sub-wash is a session leader (the 2nd lower-case `s` under STAT).

Reply:
Figured out session ID, `ps` on macOS doesn't really query it right. Wrote a small C program that checks it
```
printf("pid: %d, process groupd id: %d, session ID: %d\n", pid, getpgid(pid), getsid(pid));
```
and by default everything shares a single session.

What I've done in this PR seems to work the best, haven't come up with any variations. I'll try to wrap up all the explanation in some code comments.

Reply:
Re: https://github.com/puppetlabs/wash/pull/637#discussion_r354017719, you're saying that the `operation not permitted` error that you found on zsh happened because of something like:

* `P` and `C` share the same session, let's say `S1`
* `P` changes its process group to `B`
* `P` changes its session to `S2`
* `C` asks for `P`'s process group, which is `B`.
* `C` attempts to transfer the TTY to `B`, but that doesn't work because `P`'s session is now `S2` instead of `S1`

Reply:
Yes. Previously they were all in the same process group, now the Wash daemon is in a different process group.

I need to specifically check that this works on Linux as well, because it's getting into some low-level classification. I haven't figured out how to have an acceptance test that exercises the interactive code.

Reply:
Works on Linux.

Reply:
Ok, the new session was the only important thing. So did find a way to simplify it, and I think the explanation is clearer now too.